Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (C) 2022 Igalia S.L.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "WebKitQueryPermissionRequest.h"
+
+#include "APISecurityOrigin.h"
+#include "WebKitQueryPermissionRequestPrivate.h"
+#include "WebKitSecurityOriginPrivate.h"
+#include <wtf/glib/WTFGType.h>
+
+/**
+ * WebKitQueryPermissionRequest:
+ * @See_also: #WebKitWebView
+ *
+ * This permission request represent a user's choice to allow or deny access to "powerful features"
+ * of the platform, as specified in the [Permissions W3C
+ * Specification](https://w3c.github.io/permissions/).
+ *
+ * When signalled by the #WebKitWebView through the `query-permission` signal, the application has
+ * to respond via return value of the signal, whether it grants, denies or requests a dedicated
+ * permission prompt for the given query.
+ *
+ * When a #WebKitQueryPermissionRequest is not handled by the user, the user-agent is instructed to
+ * `prompt` the user for the given permission.
+ */
+
+struct _WebKitQueryPermissionRequest {
+    explicit _WebKitQueryPermissionRequest(const WTF::String& permissionName, API::SecurityOrigin& origin, CompletionHandler<void(std::optional<WebCore::PermissionState>)>&& completionHandler)
+        : permissionName(permissionName.utf8())
+        , completionHandler(WTFMove(completionHandler))
+    {
+        securityOrigin = API::SecurityOrigin::create(origin.securityOrigin());
+    }
+
+    ~_WebKitQueryPermissionRequest()
+    {
+        // Fallback to Prompt response unless the completion handler was already called.
+        if (completionHandler)
+            completionHandler(WebCore::PermissionState::Prompt);
+    }
+
+    CString permissionName;
+    CompletionHandler<void(std::optional<WebCore::PermissionState>)> completionHandler;
+    RefPtr<API::SecurityOrigin> securityOrigin;
```

Comment:
Could we keep a `WebKitSecurityOrigin` instead? that way the getter can be transfer none.